controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -288,52 +295,64 @@ public String getScopeName() {
+        // extract epoch from segment id.
+        // fetch epoch record for the said epoch
+        // extract segment record from it.
+        int epoch = StreamSegmentNameUtils.getEpoch(segmentId);
+        return getEpochRecord(epoch)
+                .thenApply(epochRecord -> {
+                    Optional<StreamSegmentRecord> segmentRecord = epochRecord.getSegments().stream()
+                                                                             .filter(x -> x.segmentId() == segmentId).findAny();
+                    return transform(segmentRecord
+                            .orElseThrow(() -> StoreException.create(StoreException.Type.DATA_NOT_FOUND,
+                                    "segment not found in epoch")));
+                });
     }
-
+    
     @Override
-    public CompletableFuture<List<ScaleMetadata>> getScaleMetadata() {
-        return verifyLegalState()
-                .thenCompose(v -> getHistoryIndex())
-                    .thenCompose(historyIndex -> getHistoryTable()
-                        .thenApply(historyTable -> TableHelper.getScaleMetadata(historyIndex.getData(), historyTable.getData()))
-                .thenCompose(listOfScaleRecords ->
-                        Futures.allOfWithResults(listOfScaleRecords.stream().map(record -> {
-                            long scaleTs = record.getLeft();
-                            CompletableFuture<List<Segment>> list = Futures.allOfWithResults(
-                                    record.getRight().stream().map(this::getSegment)
-                                            .collect(Collectors.toList()));
-                            return list.thenApply(segments -> new ImmutablePair<>(scaleTs, segments));
-                        }).collect(Collectors.toList())))
-                .thenApply(this::mapToScaleMetadata));
-    }
-
-    private List<ScaleMetadata> mapToScaleMetadata(List<ImmutablePair<Long, List<Segment>>> scalePair) {
-        final AtomicReference<List<Segment>> previous = new AtomicReference<>();
-        return scalePair.stream()
-                .map(pair -> {
-                    long splits = 0;
-                    long merges = 0;
-                    if (previous.get() != null) {
-                        splits = findSegmentSplitsMerges(previous.get(), pair.right);
-                        merges = findSegmentSplitsMerges(pair.right, previous.get());
-                    }
-                    previous.set(pair.getRight());
-                    return new ScaleMetadata(pair.left, pair.right, splits, merges);
-        }).collect(Collectors.toList());
+    public CompletableFuture<List<ScaleMetadata>> getScaleMetadata(final long from, final long to) {
+        // fetch history index and find epochs corresponding to "from" and "to"
+        // fetch "from epoch" from epoch record
+        // fetch epochs from history timeseries.
+        CompletableFuture<Integer> fromEpoch = findEpochAtTime(from);
```

Comment:
What are to and from here exactly? Timestamps? How are these timestamps generated? I'm actually more interested in how this is going to be used.

Reply:
yes, they are timestamp.

These are supplied by the UI. 

Earlier, we would get the entire scaleMetadata from store  (because they could be retrieved from a single store call to get history table) and then did filtering over `from` and `to` in  the rest layer.

Now we new data structures, getting the entire history is going to be an expensive operation, so we have introduced `from` and `to` into the store api and we only retrieve scale records between the desired time range.

Reply:
I don't understand what the time source is. Is this wallclock time? Who is providing the timestamps?

Reply:
Timesource has been wall clock of controller instances. 

We only ensured that epochs were created in increasing order of time but didnt bother with sychronizing wall clocks of controller. 

Though, with this changeset we have removed notion of time in regular segment's history queries, this UI specific api continues to use the notion of time. 

So the time is still recorded with our epochs, just as before, merely ensuring time is strictly in increasing order.

Reply:
The UI queries the controller to request for all scale records within a given time window (it supplies `from` and `to` parameters). 

We try to find the epoch times that fall within the requested `from` and `to` range and give it to the caller.